Src/VimWpf/Implementation/RelativeLineNumbers/Util/SafeRefreshLock.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Vim.UI.Wpf.Implementation.RelativeLineNumbers.Util
+{
+    public class SafeRefreshLock
+    {
+        private int _inProgress;
+        private int _refreshRequested;
+
+        public async Task ExecuteInLockAsync(Func<Task> operation)
+        {
+            if (!TryObtainRedrawLock())
+            {
+                return;
+            }
+
+            await operation().ConfigureAwait(false);
+
+            if (!TryReleaseRedrawLock())
+            {
+                await ExecuteInLockAsync(operation).ConfigureAwait(false);

```

Comment:
>ExecuteInLockAsync [](start = 22, length = 18)

This is an example: so in this case `operation` is executed maybe twice?

Reply:
Yes, that appears to be the intent that while we performed the operation (the side-effect free operation), another request was queued and that means, e.g., the text buffer we were using has a new snapshot or the view changed.

It does seem the async approach could be a useful strategy for all kinds of things, like the glyph margin or taggers (or even searching) that do **some** preparatory work on immutable data before interacting directly with the view on the UI thread. So I'm leaning toward keeping the async version because it only applies to a setting we didn't used to support, and will be used by a limited number of users, and so the consequences should be limited.

Reply:
Deleted.